File: docs/js/geometry.js
/**
 * Geometry utilities for the lattice triangle app.
 * 
 * Main computations (area, boundary/interior points) are done in Python via Pyodide.
 * This module provides JS utilities for validation and re-exports Python functions.
 */

import { latticeRange } from "./constants.js";
import {
  computePolygonSnapshot as pyComputeSnapshot,
  computeAdditiveSnapshot as pyComputeAdditive,
  createEmptySnapshot,
  createEmptyAdditiveSnapshot,
} from "./pyodide-loader.js";

// Re-export the Pyodide-based computation functions
export const computePolygonSnapshot = pyComputeSnapshot;
export const computeAdditiveSnapshot = pyComputeAdditive;
export { createEmptySnapshot as createEmptyPolygonSnapshot };
export { createEmptyAdditiveSnapshot };

/**
 * GCD calculation (used for JS-side validation).
 */
export function gcd(a, b) {
  let x = Math.abs(a);
  let y = Math.abs(b);
  while (y) {
    [x, y] = [y, x % y];
  }
  return x;
}

/**
 * Clamp a vertex to the lattice range.
 */
export function clampVertex(vertex) {
  return {
    x: Math.max(-latticeRange, Math.min(latticeRange, Math.round(vertex.x))),
    y: Math.max(-latticeRange, Math.min(latticeRange, Math.round(vertex.y))),
  };
}

/**
 * Calculate the signed area of a polygon (used for validation).
 */
export function polygonAreaValue(vertices) {
  if (!vertices || vertices.length < 3) return 0;
  let area2 = 0;
  for (let i = 0; i < vertices.length; i++) {
    const current = vertices[i];
    const next = vertices[(i + 1) % vertices.length];
    area2 += current.x * next.y - next.x * current.y;
  }
  return area2 / 2;
}

/**
 * Calculate orientation of three points.
 * Returns positive if counter-clockwise, negative if clockwise, 0 if collinear.
 */
export function orientation(a, b, c) {
  return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
}

/**
 * Check if point b is on segment ac.
 */
export function onSegment(a, b, c) {
  return (
    b.x >= Math.min(a.x, c.x) &&
    b.x <= Math.max(a.x, c.x) &&
    b.y >= Math.min(a.y, c.y) &&
    b.y <= Math.max(a.y, c.y)
  );
}

/**
 * Check if two line segments intersect.
 */
export function segmentsIntersect(p1, p2, p3, p4) {
  const o1 = orientation(p1, p2, p3);
  const o2 = orientation(p1, p2, p4);
  const o3 = orientation(p3, p4, p1);
  const o4 = orientation(p3, p4, p2);

  if (o1 === 0 && onSegment(p1, p3, p2)) return true;
  if (o2 === 0 && onSegment(p1, p4, p2)) return true;
  if (o3 === 0 && onSegment(p3, p1, p4)) return true;
  if (o4 === 0 && onSegment(p3, p2, p4)) return true;

  return o1 * o2 < 0 && o3 * o4 < 0;
}

/**
 * Check if a polygon is simple (no self-intersecting edges).
 */
export function isSimplePolygon(vertices) {
  if (!vertices || vertices.length < 3) return false;
  const unique = new Set(vertices.map((v) => `${v.x},${v.y}`));
  if (unique.size !== vertices.length) return false;
  if (Math.abs(polygonAreaValue(vertices)) < 0.5) return false;

  const n = vertices.length;
  for (let i = 0; i < n; i++) {
    const a1 = vertices[i];
    const a2 = vertices[(i + 1) % n];
    for (let j = i + 1; j < n; j++) {
      if (Math.abs(i - j) <= 1) continue;
      if (i === 0 && j === n - 1) continue;
      const b1 = vertices[j];
      const b2 = vertices[(j + 1) % n];
      if (segmentsIntersect(a1, a2, b1, b2)) {
        return false;
      }
    }
  }
  return true;
}

/**
 * Check if two triangles in a quadrilateral are on opposite sides of the diagonal.
 */
export function trianglesOnOppositeSides(vertices) {
  if (!vertices || vertices.length !== 4) return false;
  const [a, b, c, d] = vertices;
  const oB = orientation(a, c, b);
  const oD = orientation(a, c, d);
  if (oB === 0 || oD === 0) {
    return false;
  }
  return oB * oD < 0;
}

/**
 * Check if a point is on a line segment.
 */
export function isPointOnEdge(point, start, end) {
  const cross = (point.x - start.x) * (end.y - start.y) - (point.y - start.y) * (end.x - start.x);
  if (cross !== 0) return false;
  const minX = Math.min(start.x, end.x);
  const maxX = Math.max(start.x, end.x);
  const minY = Math.min(start.y, end.y);
  const maxY = Math.max(start.y, end.y);
  return point.x >= minX && point.x <= maxX && point.y >= minY && point.y <= maxY;
}

/**
 * Collect boundary points for a polygon (JS fallback, used by induction).
 */
export function collectBoundaryPoints(vertices) {
  const boundary = new Set();
  for (let i = 0; i < vertices.length; i++) {
    const start = vertices[i];
    const end = vertices[(i + 1) % vertices.length];
    const dx = end.x - start.x;
    const dy = end.y - start.y;
    const steps = gcd(Math.abs(dx), Math.abs(dy));
    if (steps === 0) {
      boundary.add(`${start.x},${start.y}`);
      continue;
    }
    const stepX = dx / steps;
    const stepY = dy / steps;
    for (let j = 0; j <= steps; j++) {
      boundary.add(`${start.x + stepX * j},${start.y + stepY * j}`);
    }
  }
  return Array.from(boundary)
    .map((entry) => entry.split(",").map((v) => Number(v)))
    .sort(([ax, ay], [bx, by]) => (ax === bx ? ay - by : ax - bx));
}

/**
 * Format an area value for display.
 */
export function formatArea(value) {
  if (!Number.isFinite(value)) return "0";
  return Number.isInteger(value) ? value.toString() : value.toFixed(2);
}

File: docs/py/geometry.py
"""
Lattice point geometry calculations for The Conjecture visualization.
This module is loaded via Pyodide in the browser.
"""

from math import gcd
from typing import Sequence, Tuple, List

Point = Tuple[int, int]


def _validate_points(points: Sequence[Point]) -> List[Point]:
    """Convert and validate a sequence of points."""
    pts = [tuple(int(v) for v in p) for p in points]
    if len(pts) < 3:
        raise ValueError("At least three points are required to describe a polygon.")
    return pts


def _edge_sign(point: Point, start: Point, end: Point) -> int:
    """Calculate the sign of the cross product for edge containment."""
    return (point[0] - end[0]) * (start[1] - end[1]) - (start[0] - end[0]) * (
        point[1] - end[1]
    )


def _triangle_contains(point: Point, vertices: Sequence[Point]) -> bool:
    """Check if a point is strictly inside a triangle (not on boundary)."""
    a, b, c = vertices
    d1 = _edge_sign(point, a, b)
    d2 = _edge_sign(point, b, c)
    d3 = _edge_sign(point, c, a)
    has_neg = d1 < 0 or d2 < 0 or d3 < 0
    has_pos = d1 > 0 or d2 > 0 or d3 > 0
    on_boundary = d1 == 0 or d2 == 0 or d3 == 0
    return not on_boundary and not (has_neg and has_pos)


def is_point_on_edge(point: Point, start: Point, end: Point) -> bool:
    """Check if a point lies on a line segment."""
    cross = (point[0] - start[0]) * (end[1] - start[1]) - (point[1] - start[1]) * (
        end[0] - start[0]
    )
    if cross != 0:
        return False
    min_x, max_x = sorted((start[0], end[0]))
    min_y, max_y = sorted((start[1], end[1]))
    return min_x <= point[0] <= max_x and min_y <= point[1] <= max_y


def _is_point_in_polygon(point: Point, vertices: Sequence[Point]) -> bool:
    """Check if a point is strictly inside a polygon using ray casting."""
    pts = _validate_points(vertices)
    x, y = point
    inside = False
    n = len(pts)
    for i in range(n):
        start = pts[i]
        end = pts[(i + 1) % n]
        if is_point_on_edge(point, start, end):
            return False
        x1, y1 = start
        x2, y2 = end
        if (y1 > y) != (y2 > y):
            denom = y2 - y1
            if denom != 0:
                xinters = (x2 - x1) * (y - y1) / denom + x1
                if xinters > x:
                    inside = not inside
    return inside


def _point_in_shape(point: Point, vertices: Sequence[Point]) -> bool:
    """Check if a point is strictly inside a shape (triangle or polygon)."""
    pts = _validate_points(vertices)
    if len(pts) == 3:
        return _triangle_contains(point, pts)
    return _is_point_in_polygon(point, pts)


def get_polygon_area(points: Sequence[Point]) -> float:
    """Calculate the area of a polygon using the shoelace formula."""
    pts = _validate_points(points)
    area2 = 0
    for i in range(len(pts)):
        x1, y1 = pts[i]
        x2, y2 = pts[(i + 1) % len(pts)]
        area2 += x1 * y2 - x2 * y1
    return abs(area2) / 2.0


def get_boundary_points(points: Sequence[Point]) -> int:
    """Count the number of lattice points on the polygon boundary."""
    pts = _validate_points(points)
    boundary = 0
    for i in range(len(pts)):
        x1, y1 = pts[i]
        x2, y2 = pts[(i + 1) % len(pts)]
        boundary += gcd(abs(x2 - x1), abs(y2 - y1))
    return boundary


def get_interior_points(points: Sequence[Point]) -> int:
    """Count the number of lattice points strictly inside the polygon."""
    pts = _validate_points(points)
    min_x = min(p[0] for p in pts)
    max_x = max(p[0] for p in pts)
    min_y = min(p[1] for p in pts)
    max_y = max(p[1] for p in pts)

    interior = 0
    n = len(pts)
    for x in range(min_x, max_x + 1):
        for y in range(min_y, max_y + 1):
            candidate = (x, y)
            if candidate in pts:
                continue
            if any(is_point_on_edge(candidate, pts[i], pts[(i + 1) % n]) for i in range(n)):
                continue
            if _point_in_shape(candidate, pts):
                interior += 1
    return interior


def collect_boundary_points(vertices: Sequence[Point]) -> List[Point]:
    """Collect all lattice points on the polygon boundary."""
    pts = _validate_points(vertices)
    boundary = set()
    for i in range(len(pts)):
        start = pts[i]
        end = pts[(i + 1) % len(pts)]
        dx, dy = end[0] - start[0], end[1] - start[1]
        steps = gcd(abs(dx), abs(dy))
        if steps == 0:
            boundary.add(start)
            continue
        step_x, step_y = dx // steps, dy // steps
        for j in range(steps + 1):
            boundary.add((start[0] + step_x * j, start[1] + step_y * j))
    return sorted(boundary)


def collect_interior_points(vertices: Sequence[Point]) -> List[Point]:
    """Collect all lattice points strictly inside the polygon."""
    pts = _validate_points(vertices)
    min_x = min(p[0] for p in pts)
    max_x = max(p[0] for p in pts)
    min_y = min(p[1] for p in pts)
    max_y = max(p[1] for p in pts)

    interior = []
    n = len(pts)
    for x in range(min_x, max_x + 1):
        for y in range(min_y, max_y + 1):
            candidate = (x, y)
            if candidate in pts:
                continue
            if any(is_point_on_edge(candidate, pts[i], pts[(i + 1) % n]) for i in range(n)):
                continue
            if _point_in_shape(candidate, pts):
                interior.append(candidate)
    return interior


def compute_snapshot(vertices: Sequence[Point]) -> dict:
    """
    Compute a complete snapshot of polygon statistics.
    
    Returns a dictionary with:
    - area: The polygon area
    - boundary: Number of boundary lattice points
    - interior: Number of interior lattice points
    - boundary_points: List of boundary point coordinates
    - interior_points: List of interior point coordinates
    """
    pts = [tuple(int(v) for v in pair) for pair in vertices]
    return {
        "area": get_polygon_area(pts),
        "boundary": get_boundary_points(pts),
        "interior": get_interior_points(pts),
        "boundary_points": collect_boundary_points(pts),
        "interior_points": collect_interior_points(pts),
    }


def _segment_interior_lattice_points(start: Point, end: Point) -> Tuple[int, List[Point]]:
    """Get interior lattice points on a line segment (excluding endpoints)."""
    dx, dy = end[0] - start[0], end[1] - start[1]
    steps = gcd(abs(dx), abs(dy))
    if steps <= 1:
        return 0, []
    step_x, step_y = dx // steps, dy // steps
    points = [
        (start[0] + step_x * i, start[1] + step_y * i)
        for i in range(1, steps)
    ]
    return steps - 1, points


def compute_additive_snapshot(vertices: Sequence[Point]) -> dict:
    """
    Compute snapshot for two triangles sharing an edge (quadrilateral).
    
    Expects exactly 4 vertices forming a convex quadrilateral where
    vertices[0] and vertices[2] define the shared diagonal.
    
    Returns a dictionary with:
    - t1: Snapshot for triangle (v0, v1, v2)
    - t2: Snapshot for triangle (v0, v3, v2)
    - union: Snapshot for the full quadrilateral
    - shared_edge: Info about the shared diagonal
    """
    pts = _validate_points(vertices)
    if len(pts) != 4:
        raise ValueError("Additive snapshot expects exactly four vertices.")
    a, b, c, d = pts

    triangle_one = compute_snapshot((a, b, c))
    triangle_two = compute_snapshot((a, d, c))
    union = compute_snapshot(pts)

    edge_interior_count, edge_interior_points = _segment_interior_lattice_points(a, c)

    return {
        "t1": triangle_one,
        "t2": triangle_two,
        "union": union,
        "shared_edge": {
            "points": edge_interior_points,
            "interior_count": edge_interior_count,
            "endpoints": [a, c]
        },
    }

